test/meta_transaction_test.ts

```diff
@@ -0,0 +1,742 @@
+                expect(validationResults.rejected.length, 'mesh should not reject any orders').to.be.eq(0);
+                const largeBuyAmount = MAKER_ASSET_AMOUNT.times(2).toString();
+                const route = constructRoute({
+                    baseRoute: `${META_TRANSACTION_PATH}/quote`,
+                    queryParams: {
+                        ...DEFAULT_QUERY_PARAMS,
+                        buyAmount: largeBuyAmount,
+                        takerAddress,
+                    },
+                });
+                const response = await httpGetAsync({ route });
+                expect(response.type).to.be.eq('application/json');
+                expect(response.status).to.be.eq(HttpStatus.OK);
+                assertCorrectQuote({
+                    quote: response.body,
+                    expectedBuyAmount: largeBuyAmount,
+                    expectedOrders: validationResults.accepted.map(accepted => accepted.signedOrder),
+                    expectedPrice: '1.5',
+                });
+            });
+        });
+    });
+
+    describe('/submit tests', () => {
+        const requestBase = `${META_TRANSACTION_PATH}/submit`;
+
+        context('failure tests', () => {
+            it('should return InvalidAPIKey error if invalid UUID supplied as API Key', async () => {
+                const response = await httpPostAsync({ route: requestBase, headers: { '0x-api-key': 'foobar' } });
+                expect(response.status).to.be.eq(HttpStatus.BAD_REQUEST);
+                expect(response.type).to.be.eq('application/json');
+                expect(response.body).to.be.deep.eq({
+                    code: GeneralErrorCodes.InvalidAPIKey,
+                    reason: generalErrorCodeToReason[GeneralErrorCodes.InvalidAPIKey],
+                });
+            });
+        });
+
+        context('success tests', () => {
+            let meshUtils: MeshTestUtils;
+
+            function signZeroExTransaction(transaction: ZeroExTransaction, signingAddress: string): string {
+                const transactionHashBuffer = transactionHashUtils.getTransactionHashBuffer(transaction);
+                const pkIdx = accounts.indexOf(signingAddress);
+                expect(pkIdx, 'signing address is invalid').to.be.gte(0);
+                const privateKey = constants.TESTRPC_PRIVATE_KEYS[pkIdx];
+                return `0x${signingUtils
+                    .signMessage(transactionHashBuffer, privateKey, SignatureType.EthSign)
+                    .toString('hex')}`;
+            }
+
+            describe('single order submission', () => {
+                let validationResults: ValidationResults;
+                const price = '1';
+                const sellAmount = calculateSellAmount(buyAmount, price);
+
+                // NOTE(jalextowle): This must be a `before` hook because `beforeEach`
+                // hooks execute after all of the `before` hooks (even if they are nested).
```

Comment:
I noticed the earlier test cases use `beforeEach`. Seems like setting up mesh and creating a blockchain checkpoint *after* all the `before` hooks would be the desired behaviour? 

Couple qs, 1) what's different about this case vs earlier cases; and 2) do these blocks execute in scope order right now? e.g. top level scope `before`, then nested `before`?

Reply:
This is supposed to test calling `/price` then `/quote` then `/submit`. If we use the `beforeEach` pattern, then this would revert all of the progress that had been made during previous tests and wouldn't really be as much of an `end-to-end test`.